raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -167,16 +229,35 @@ public boolean updateLocalEndOffset(LogOffsetMetadata endOffsetMetadata) {
 
     private static class ReplicaState implements Comparable<ReplicaState> {
         final int nodeId;
-        boolean hasEndorsedLeader;
+
         Optional<LogOffsetMetadata> endOffset;
+
         OptionalLong lastFetchTimestamp;
+        long lastCaughtUpTimeMs = 0L;
+        long lastFetchLeaderLogEndOffset;
 
         public ReplicaState(int nodeId,
-                            boolean hasEndorsedLeader) {
+                            long lastLeaderLogEndOffset) {
             this.nodeId = nodeId;
-            this.hasEndorsedLeader = hasEndorsedLeader;
             this.endOffset = Optional.empty();
+
             this.lastFetchTimestamp = OptionalLong.empty();
+            this.lastFetchLeaderLogEndOffset = lastLeaderLogEndOffset;
+        }
+
+        void updateFetchState(long currentFetchOffset,
+                              long currentLeaderEndOffset,
+                              long currentFetchTimeMs) {
+            if (currentFetchOffset >= currentLeaderEndOffset) {
+                lastCaughtUpTimeMs = currentFetchTimeMs;
```

Comment:
In the KIP our definition for lastCaughtUpTimeMs is regarding "replica was caught up to the high watermark", while here it's defined by leader end offset, is that intentional?

Also I think if we compare the fetch offset with hw indeed, then we do not need to remember the `lastFetchTimestamp` to give it a second chance if it has by-passed the previous log end offset; instead we always compare against the newest hw.

Reply:
That's right, I was just porting from the Cluster Replica definition for lastCaughtUpTime. Maybe @hachikuji could also share some thoughts on whether we care more about HW or the end offset.